src/error.rs

```diff
@@ -34,12 +35,61 @@ struct SerializeError<'a> {
 }
 
 impl Error {
-    /// Returns a new [`Error`] with the provided message.
+    /// Create a new [`Error`] from the provided source.
     ///
     pub fn new(source: DynError) -> Self {
         Self::internal_server_error(source)
     }
 
+    /// Create a new [`Error`] from the provided [`io::Error`]. The status code
+    /// of the error returned will correspond to `source.kind()`.
+    ///
+    pub fn from_io(source: io::Error) -> Self {
+        match source.kind() {
```

Comment:
There shouldn't be an issue allocating in a match branch. My reason for mentioning it is in case someone knows otherwise.